design/open/lakefs-on-kv-testing-plan.md

```diff
@@ -0,0 +1,133 @@
+
+### Performance
+* **A benchmark for the migration process itself:**
+  * How do we load the DB with data to migrate for benchmarking?
+  * Ideally, run on several data sizes (tens / thousands / millions branches & commits, various numbers of users, etc.) to get sense of how scaling will affect
+  * What is the expected/acceptable outcome?
+  * Should be built gradually, as we support and migrate additional package
+
+* **Benchmark comparison of both DBs**
+  * Compare performance before and after migration - does not include the migration itself
+  * Define package specific set of operation to benchmark
+  * Run with both Table DB and KV Store and compare the results
+  * What is the accepted degradation, if any?
+  * Need to consider various scales
+## Per Package DB Testing
+
+### ```pkg/gateway/multiparts```
+* DB is used to track the start and end of a multipart upload. All DB accesses are done via `mutltiparts.Tracker`. Entries are created once, read-accesses multiple times and deleted upon completion
+* Currently unit tests cover correctness of DB accesses in both good and error paths.
+* 83.8% coverage
+* Currently system tests cover a simple multipart upload - a single object with 7 parts, good path only
+* What is missing:
+  * Performance - Can that sustain heavy loads: multi-multipart uploads with a lot of parts
+  * Concurrency, that is derived from the previous bullet (can we increase concurrency by using smaller parts?)
+  * Migration - Verify that data migrated from Table to KV, in the middle of multipart upload, is still usable
+
+
+### ```pkg/actions```
+* DB is used to store actions runs and hook runs with results. Db is read by the actions service to handle requests from the various clients. For each run there is a new entry in `actions_run` table and an entry for each hook in `actions_run_hooks` table.
+Updates are done for commitID, as part of `post` hooks
+* Currently unit tests cover hooks run and verify the results in DB using the service reading functions. There are also error testings (hook failures)
+* 67.6% coverage
+* What is missing:
+  * DB Failure during hook action (TBD - what is the expected behavior? Is that interesting to test?)
+  * Performance under load - not sure a load on actions/hooks is relevant
+  * Concurrent actions/hooks execution, which derives concurrent DB accesses. This gets more interesting if load is involved (in case load is relevant)
+
+### ```pkg/auth```
+* DB is used to store installation metadata, users and policies
+* Only 6.8% coverage
+* Some benchmarks exists that covers effectivePolicies (multi-table JOIN)
+* What is missing:
+  * Need to increase code coverage
+  * Comparison benchmark to toggle between Tables and KV and verify there is no degradation
+  * Migration tests:
+    * verify that data is migrated correctly
+    * Authentication transactions in during migration (data that was written to tables is read from KV)
```

Comment:
Can you explain how transactions should occur during migration? Migration should be done when lakeFS is still down

Reply:
`During migration` stands for `before and after migration`, rather than during the actual process of the migration. What I meant is that data can sustain the migration process, e.g. a `multipart upload` that is started and written to the `gateway_multiparts` table, can resume and complete after the migration to KV Store